File: 2020/day2.py
# day2
def is_valid_password(x):
    least = int(x[0])
    most  = int(x[1])
    char  = x[2]
    pwd   = x[3]

    num = 0
    for c in pwd:
        if c == char:
            num += 1
    if num >= least and num <= most:
        return True
    return False


def is_valid_password2(x):
    i = int(x[0]) - 1
    j = int(x[1]) - 1
    char  = x[2]
    pwd   = x[3]

    if pwd[i] != pwd[j]:
        if pwd[i] == char or pwd[j] == char:
            return True
    return False


count  = 0
count2 = 0

f = open('day2.txt', 'r')
for x in f:
    x = x.replace('-', ' ')
    x = x.replace(':', '')
    x = x.split()
    if is_valid_password(x):
        count += 1
    if is_valid_password2(x):
        count2 += 1
        
print('valid passwords part 1:', count)
print('valid passwords part 2:', count2)

File: 2020/day2.java

import java.io.File; 
import java.util.*; 

public class day2 {
	public static void main(String[] args) {
		int count  = 0;
		int count2 = 0;
		
		ArrayList<String> data = read_file("day2.txt");
		for (int i = 0; i < data.size(); i++) {
			String whole = data.get(i);
			String[] split = whole.split(" |-|: "); 

			// part 1
			if (is_valid_pwd(split)) {
				count++;
			}
			// part 2
			if (is_valid_pwd2(split)) {
				count2++;
			}
		}
		
		System.out.println("Valid Passwords: " + count);	// part 1
		System.out.println("Valid Passwords: " + count2);	// part 2
		
	}

	private static boolean is_valid_pwd2(String[] s) {
		int i      = Integer.parseInt(s[0]) - 1;
		int j      = Integer.parseInt(s[1]) - 1;
		char ch    = s[2].charAt(0);
		String pwd = s[3];
		
		if (pwd.charAt(i) != pwd.charAt(j)) {
			if (pwd.charAt(i) == ch || pwd.charAt(j) == ch) {
				return true;
			}
		}
		return false;
	}

	private static boolean is_valid_pwd(String[] s) {
		int least  = Integer.parseInt(s[0]);
		int most   = Integer.parseInt(s[1]);
		char ch    = s[2].charAt(0);
		String pwd = s[3];
		int num    = 0;
		
		for (char c: pwd.toCharArray()) {
			if (c == ch){
				num++;
			}
		}
		if (num >= least && num <= most) {
			return true;
		}
		return false;
	}

	private static ArrayList<String> read_file(String fileName) {
		ArrayList<String> list = new ArrayList<String>();
	    
	    try {
	    	Scanner scan = new Scanner(new File(fileName));
	    	while(scan.hasNext()){
	    		list.add(scan.nextLine());
	    	}
	    }
	    catch(Exception e) {	
	    }
	    
		return list;
	}
	
}
